File: haskell/RefTrans.hs
module Main
where

f :: IO a -> IO a -> IO a
f ioaction1 ioaction2 = do
    ioaction1
    ioaction2

main :: IO ()
main = do
  putStrLn "----- Referential Transparency"
  f (putStrLn "hi") (putStrLn "hi")
  putStrLn "-----"
  -- is equivalent to
  let x = putStrLn "hi" in f x x
  putStrLn "-----"

File: src/main/scala/ioMonadForCats/reftrans/RefTrans.scala
package ioMonadForCats.reftrans

/*
  see: https://typelevel.org/blog/2017/05/02/io-monad-for-cats.html
 */
object RefTrans extends App {

  println("\n-----")

  def f(ioaction1: Unit, ioaction2: Unit): Unit = {
    ioaction1
    ioaction2
  }

  f(println("hi"), println("hi"))
  println("-----")
  // isn't really equivalent to!
  val x: Unit = println("hi")
  f(x, x)

  println("-----\n")
}
